### PR TITLE
[pytx] mypy supportand fixes

### PR DESCRIPTION
**Pull request by BarrettOlson**
Summary
---------

This is the remaining pieces from break up #873 and is mostly just around fixing types.

I would recommend reviewing commit by commit. 

### [`aec21f10`](https://github.com/facebook/ThreatExchange/pull/878/commits/aec21f10e2778c1b18a7c23ac7fff0c0b08af3c5) add simple type hints
- no logic changes
### [`304b9340`](https://github.com/facebook/ThreatExchange/pull/878/commits/304b9340b9bfa077b6858e30290f83ba57407048) add typing and simple fixes to threatexchange cli commands
- minor fixes + typing
### [`34a63fce`](https://github.com/facebook/ThreatExchange/pull/878/commits/34a63fcebb2d916280d9d52ed7dc76cbf65743fa) adds typing, ignores, list-to-set, and null checks on threat_update utils and descriptor.py
- type + some logic fixes / casting

### [`e7221b8c`](https://github.com/facebook/ThreatExchange/pull/878/commits/e7221b8cc8bdb5eead51bb59b978ee1b7c926905) update type hints and add set to list casts for serialization classes
- threat_update specific typing

### [`da27f81d`](https://github.com/facebook/ThreatExchange/pull/878/commits/da27f81db20b6707795afe2c400e8c4645c8449d) add mypy checks to CLI
- adds stuff to github actions

### [`06afe107`](https://github.com/facebook/ThreatExchange/pull/878/commits/06afe1073cf0dcb4d627271688130d64e2064e20) final typing fixes in pdq_index and faiss files
- typing fixes needed from files in #875 

### [`0ecc14c5`](https://github.com/facebook/ThreatExchange/pull/878/commits/0ecc14c563616ccc1b3863614cfe5942bb535b5e) add typestubs to make ci happy

Test Plan
---------

Repeat of #873 testing after rebase

### CODE REVIEW

File: python-threatexchange/threatexchange/cli/fetch.py [hunk 1]
```diff
@@ -74,7 +74,7 @@ def __init__(
         # Print first update after 5 seconds
         self.last_update_printed = time.time() - self.PROGRESS_PRINT_INTERVAL_SEC + 5
         self.processed = 0
-        self.counts = collections.Counter()
+        self.counts: t.Dict[str, int] = collections.Counter()
```
Comment (schatten) [inline python-threatexchange/threatexchange/cli/fetch.py:77]: Always thought inferred typing for stdlib would take care of this.

File: python-threatexchange/threatexchange/cli/hash_cmd.py [hunk 2]
```diff
@@ -103,7 +103,7 @@ def execute(self, api: ThreatExchangeAPI, dataset: Dataset) -> None:
 
         for inp in self.input_generator:
             hash_fn = lambda s, t: s.hash_from_file(t)
-            signal_types = file_hashers
+            signal_types: t.List[t.Any] = file_hashers
```
Comment (schatten) [inline python-threatexchange/threatexchange/cli/hash_cmd.py:106]: nit: Can do with more typing at line 101.

File: python-threatexchange/threatexchange/cli/match.py [hunk 3]
```diff
@@ -141,7 +141,7 @@ def interpret_token(
                     no_stderr=True,
                 )
             else:
-                yield parsed
+                yield from parsed
```
Comment (schatten) [inline python-threatexchange/threatexchange/cli/match.py:144]: Woah! Nicely caught and nicely done.!

File: python-threatexchange/threatexchange/cli/match.py [hunk 4]
```diff
@@ -155,7 +155,9 @@ def execute(self, api: ThreatExchangeAPI, dataset: Dataset) -> None:
         )
 
         file_matchers = [s for s in all_signal_types if isinstance(s, FileMatcher)]
-        str_matchers = [s for s in all_signal_types if isinstance(s, StrMatcher)]
+        str_matchers: t.List[t.Any] = [
```
Comment (schatten) [inline python-threatexchange/threatexchange/cli/match.py:158]: sounds like all str_matchers will be StrMatcher instances. To note t.List is an [invariant type](https://mypy.readthedocs.io/en/stable/common_issues.html#invariance-vs-covariance). If you are having trouble with typing this, use `t.Sequence[StrMatcher]`.

File: python-threatexchange/threatexchange/collab_config.py [hunk 5]
```diff
@@ -68,7 +68,7 @@ def store(self, filename: str):
             json.dump(out, f, indent=4)
 
     @classmethod
-    def get_example_config(cls) -> "CollabConfig":
+    def get_example_config(cls) -> "CollaborationConfig":
```
Comment (schatten) [inline python-threatexchange/threatexchange/collab_config.py:71]: Pretty sure we can get  PEP running just based on this line. 🤣

File: python-threatexchange/threatexchange/common.py [hunk 6]
```diff
@@ -72,6 +72,6 @@ def normalize_url(url: str) -> str:
     # https://www.facebook.com => www.facebook.com
     url = parsed.geturl().replace(scheme, "", 1)
     # Ensure URL is utf-8 encoded
-    url = url.encode("utf-8")
+    encoded_url = url.encode("utf-8")
```
Comment (schatten) [inline python-threatexchange/threatexchange/common.py:75]: nit: Just return here?

File: python-threatexchange/threatexchange/dataset.py [hunk 7]
```diff
@@ -129,4 +129,4 @@ def load_index(
         if not path.exists():
             return None
         with path.open("rb") as fin:
-            return signal_type.get_index_cls.deserialize(fin)
+            return signal_type.get_index_cls().deserialize(fin)
```
Comment (schatten) [inline python-threatexchange/threatexchange/dataset.py:132]: typing caught an actual error!

File: python-threatexchange/threatexchange/signal_type/pdq_index.py [hunk 8]
```diff
@@ -80,11 +80,11 @@ def serialize(self, fout: t.BinaryIO) -> None:
...
 
     @classmethod
-    def deserialize(cls, fin: t.BinaryIO) -> "SignalTypeIndex[T]":
+    def deserialize(cls, fin: t.BinaryIO) -> "SignalTypeIndex[IndexT]":
         """
-        Instanciate an index from a previous call to serialize
+        Instantiate an index from a previous call to serialize
         """
-        return pickle.loads(fin)
+        return pickle.loads(fin.read())
```
Comment (schatten) [inline python-threatexchange/threatexchange/signal_type/pdq_index.py:87]: Woah! How did this work prior to this fix?

File: python-threatexchange/threatexchange/signal_type/signal_base.py [hunk 9]
```diff
@@ -214,7 +214,7 @@ class SimpleSignalType(SignalType, HashMatcher):
     Assumes that the signal type can easily merge on a string.
     """
 
-    INDICATOR_TYPE: t.Union[str, t.Tuple[str]] = ()
+    INDICATOR_TYPE: t.Union[str, t.Tuple[str, ...]] = ()
```
Comment (schatten) [inline python-threatexchange/threatexchange/signal_type/signal_base.py:217]: TIL about the ellipsis syntax in mypy!

File: python-threatexchange/threatexchange/signal_type/signal_base.py [hunk 10]
```diff
@@ -103,7 +103,7 @@ def compare_hash(cls, hash1: str, hash2: str) -> int:
 
     ##########################################################################
     # TODO - Remove  these methods after refactor
-    def process_descriptor(self, descriptor: t.Dict[str, t.Any]) -> bool:
+    def process_descriptor(self, descriptor) -> bool:
```
Comment (schatten) [inline python-threatexchange/threatexchange/signal_type/signal_base.py:106]: Why not use `ThreatDescriptor` here?

File: python-threatexchange/threatexchange/cli/fetch.py (hunk 1 above)
Comment (BarrettOlson) [reply python-threatexchange/threatexchange/cli/fetch.py:77]: me too but mypy didn't agree

**Review by schatten — APPROVED** [review]
LGTM. Some questions inline. Nothing "BLOCKING". Trust your judgement on this.

File: python-threatexchange/threatexchange/cli/match.py (hunk 4 above)
Comment (BarrettOlson) [reply python-threatexchange/threatexchange/cli/match.py:158]: I tried but both using `t.Sequence` and `t.Sequence[t.Union[...]]` fall apart on line 173 of this file and I don't want to change too much of the logic in one PR.

File: python-threatexchange/threatexchange/cli/hash_cmd.py (hunk 2 above)
Comment (BarrettOlson) [reply python-threatexchange/threatexchange/cli/hash_cmd.py:106]: L101's type: `List[Type[<subclass of "SignalType" and "FileHasher">]]` I couldn't find a clean way to type that up there so wanted to minimize the amount of `t.Any` usage gets in the scope by adding the type hint later

File: python-threatexchange/threatexchange/common.py (hunk 6 above)
Comment (BarrettOlson) [reply python-threatexchange/threatexchange/common.py:75]: sure!

File: python-threatexchange/threatexchange/signal_type/pdq_index.py (hunk 8 above)
Comment (BarrettOlson) [reply python-threatexchange/threatexchange/signal_type/pdq_index.py:87]: I didn't we just don't use it. 
https://github.com/facebook/ThreatExchange/blob/main/hasher-matcher-actioner/hmalib/indexers/s3_indexers.py#L48-L55

**Pull request merged**